In [1]:
from scipy.optimize import leastsq
from scipy.optimize import curve_fit
import matplotlib.pyplot as plt
import numpy as np

#Revenues from The Doctor Movie:
t= np.linspace(1.0, 5, num=5) # timeline
revenues=np.array([0, 0.10,3,5.2,7]) # revenues in $ millions of dollars
c_revenues=np.cumsum(revenues) # cumulative revenues in $ millions of dollars

# initial variables(M, P, Q) as estimates (trying several variables)
vars = [34.85, 0.07,0.49]

# Min Squared Error - Bass Model
def Model (vars, t, sales):
    M = vars[0]
    P = vars[1]
    Q = vars[2]
    for i in range (len (c_revenues)):
        Bass= (P+(Q *(c_revenues[i])/M))*(M - c_revenues[i])
    return (Bass - (revenues))
          
# non linear least square fitting
varfinal,success = leastsq(Model, vars, args=(t, revenues))

# Estimation of M, P, Q
m = varfinal[0]
p = varfinal[1]
q = varfinal[2]
print(varfinal)

[ 27.32595273   0.08723368   0.29864641]


The parameters to estimate are the following:								
m = number of people estimated to eventually watch the movie								
q = coefficient of imitation (based on word-of-mouth influences)								
p = coefficient of innovation (based on watching the movie without any influence from others. 								
								
The non-linear optimization model consists of minimizing the squared errors (E_t) in which the error is the difference between the forecaseted and actual number of adopters for a specific time period t. 

We have obtain the following estimates: 
m = 27.3259
p = 0.08723
q = 0.29864